crates/typst/src/layout/grid/layout.rs

```diff
@@ -278,6 +373,114 @@ impl CellGrid {
     }
 }
 
+/// Given a cell's requested x and y, the vector with the resolved cell
+/// positions, the `auto_index` counter (determines the position of the next
+/// `(auto, auto)` cell) and the amount of columns in the grid, returns the
+/// final index of this cell in the vector of resolved cells.
+fn resolve_cell_position(
+    cell_x: Smart<usize>,
+    cell_y: Smart<usize>,
+    resolved_cells: &[Option<Cell>],
+    auto_index: &mut usize,
+    columns: usize,
+) -> HintedStrResult<usize> {
+    // Translates a (x, y) position to the equivalent index in the final cell vector.
+    // Errors if the position would be too large.
+    let cell_index = |x, y: usize| {
+        y.checked_mul(columns)
+            .and_then(|row_index| row_index.checked_add(x))
+            .ok_or_else(|| {
+                let position = match (cell_x, cell_y) {
+                    (Smart::Auto, Smart::Auto) => EcoString::default(),
+                    (Smart::Custom(x), Smart::Auto) => eco_format!("at column {x} "),
+                    (Smart::Auto, Smart::Custom(y)) => eco_format!("at row {y} "),
+                    (Smart::Custom(x), Smart::Custom(y)) => {
+                        eco_format!("at column {x}, row {y} ")
+                    }
+                };
+                HintedString::from(eco_format!(
+                    "the position of a cell {position}would be too large"
+                ))
+            })
+    };
+    match (cell_x, cell_y) {
+        // Fully automatic cell positioning. The cell did not
+        // request a coordinate.
+        (Smart::Auto, Smart::Auto) => {
+            // Let's find the first available position starting from the
+            // automatic position counter, searching in row-major order.
+            // We have to sum 'next_auto_index' at the end since we'd have
+            // the position after the initial skip, not from the start of the
+            // vector.
+            let next_auto_index = *auto_index;
+            let resolved_index = resolved_cells
+                .iter()
+                .skip(next_auto_index)
+                .position(Option::is_none)
+                .map(|i| i + next_auto_index)
+                .unwrap_or_else(|| resolved_cells.len());
+
+            // Ensure the next cell with automatic position will be
+            // placed after this one (maybe not immediately after).
+            *auto_index = resolved_index + 1;
+
+            Ok(resolved_index)
+        }
+        // Cell has chosen its exact position.
+        (Smart::Custom(cell_x), Smart::Custom(cell_y)) => {
+            if cell_x >= columns {
+                return Err(HintedString::from(eco_format!(
+                    "a cell could not be placed at invalid column {cell_x}"
```

Comment:
if the errors at the cell's span, we could the probably drop the `"a "` and just says `"cell could not ..."`.